### PR TITLE
Feature/http refactoring

### PR DESCRIPTION
**Pull request by Tigraine**
Refactored out the license call to be promise based and removed all http calls from Analytics.ts

### CODE REVIEW

File: js/core/Analytics.ts [hunk 1]
```diff
@@ -12,35 +10,36 @@ import {AnalyticsStateMachine} from '../types/AnalyticsStateMachine';
...
   static LICENSE_CALL_PENDING_TIMEOUT = 200;
   static PAGE_LOAD_TYPE_TIMEOUT = 200;
   static CAST_RECEIVER_CONFIG_MESSAGE = 'CAST_RECEIVER_CONFIG_MESSAGE';
 
   private config: AnalyicsConfig;
-  private licenseCall: LicenseCall;
-  private analyticsCall: AnalyticsCall;
+  private backend: Backend
   private droppedSampleFrames: number;
   private licensing: string;
```
Comment (peterpf) [inline js/core/Analytics.ts]: `licensing` can be removed

File: js/core/Analytics.ts [hunk 2]
```diff
@@ -12,35 +10,36 @@ import {AnalyticsStateMachine} from '../types/AnalyticsStateMachine';
...
   constructor(config: AnalyicsConfig) {
     this.config = config;
-    this.licenseCall = new LicenseCall();
-    this.analyticsCall = new AnalyticsCall();
+
+    const domain = Utils.sanitizePath(window.location.hostname);
+    this.backend = new LicenseCheckingBackend({ key: config.key, domain, version: analyticsVersion });
     this.sample = {};
     this.droppedSampleFrames = 0;
     this.licensing = 'waiting';
```
Comment (peterpf) [inline js/core/Analytics.ts]: `licensing` can be removed

File: js/core/Backend.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
+import {Sample} from '../types/Sample';
+import {LicensingRequest, LicensingResponse, LicensingResult} from '../types/LicensingRequest';
```
Comment (peterpf) [inline js/core/Backend.ts]: `LicensingResponse` can be removed

File: tests/Backend.test.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    message: 'here you go'
+}
+jest.mock('../js/utils/LicenseCall', () => {
+    return {
+        LicenseCall: jest.fn().mockResolvedValue(granted)
+    }
+});
+
+test('constructor sends licensing request', () => {
+    const backend = new Backend({
```
Comment (peterpf) [inline tests/Backend.test.ts]: Should `Backend` be `LicenseCheckingBackend` ?

File: js/core/Backend.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    this.performLicenseCheck(info);
+  }
+
+  async performLicenseCheck(info: LicensingRequest) {
+    try {
+      const result = await LicenseCall(info.key, info.domain, info.version);
+      if (result.status === LicensingResult.Granted) {
+        const remoteBackend = new RemoteBackend();
+        (this.backend as QueueBackend).flushTo(remoteBackend);
+        this.backend = remoteBackend;
```
Comment (rolandgriesser) [inline js/core/Backend.ts]: Are we sure that will always work?
Let's say we have 10 items in the queue, we flush them, which takes some time (forEach isn't awaiting, but for whatever reason, the code in the forEach takes a bit longer) - and we get a new sample in the meantime. Will the .forEach in the flush take this new sample into account, or will it get lost? maybe we should set the backend to remoteBackend first and then flush the queueBackend - worst case the events won't get to the server in the wrong order, which isn't really a problem

Comment (Tigraine) [reply js/core/Backend.ts]: Discussed in person - JS is non-blocking single-threaded so there is no way to have concurrency between `forEach` draining the list and the backend being re-assigned

**Pull request closed without merging**